### PR TITLE
type/compatibility: fix ambiguous error in `insert... select... on du…

### PR DESCRIPTION
**Pull request by jingyugao**
…plicatted key`

Signed-off-by: jingyugao <1121087373@qq.com>

### What problem does this PR solve? 
fix https://github.com/pingcap/tidb/issues/11024

### What is changed and how it works?
When FindColumnAndIndex, if there are two colums match the target column name and the two column is from insertSelect, dont return an error.

### Check List 

Tests 

 Unit test and  Integration test

Code changes

Just add some if-else and unexported field.

Side effects

None

Related changes

None

### CODE REVIEW

File: expression/schema.go [hunk 1]
```diff
@@ -107,6 +107,9 @@ func (s *Schema) FindColumnAndIndex(astCol *ast.ColumnName) (*Column, int, error
 			if idx == -1 {
 				idx = i
 			} else {
+				if col.IsInsertSelect && s.Columns[idx].IsInsertSelect {
```
Comment (eurekaka) [inline expression/schema.go]: This looks not right, check out this case:
In MySQL:
```
mysql> create table t2(f1 int primary key, f2 int);
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t2 values(1,2),(2,1);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> INSERT INTO t2 (f1, f2) SELECT f2 f1, f1 FROM t2 src ON DUPLICATE KEY UPDATE f1 = 10 + src.f1;
Query OK, 4 rows affected (0.00 sec)
Records: 2  Duplicates: 2  Warnings: 0

mysql> select * from t2;
+----+------+
| f1 | f2   |
+----+------+
| 11 |    1 |
| 12 |    2 |
+----+------+
```
while in TiDB after applying your patch:
```
mysql> create table t2(f1 int primary key, f2 int);
Query OK, 0 rows affected (0.00 sec)

mysql> insert into t2 values(1,2),(2,1);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> INSERT INTO t2 (f1, f2) SELECT f2 f1, f1 FROM t2 src ON DUPLICATE KEY UPDATE f1 = 10 + src.f1;
Query OK, 4 rows affected (0.00 sec)
Records: 2  Duplicates: 2  Warnings: 0

mysql> select * from t2;
+----+------+
| f1 | f2   |
+----+------+
| 11 |    2 |
| 12 |    1 |
+----+------+
```

Seems like MySQL would resolve the `src.f1` to the column which has original name `f1` in this query, not the column which has alias name `f1` while its original column name is `f2`.

Comment (jingyugao) [reply expression/schema.go]: this seems anther bug.
```sql
 drop table if exists t2;
         create table t2(f1 int primary key, f2 int);
         insert into t2 values(1,2),(2,1);
INSERT INTO t2 (f1, f2) SELECT f2 f1, f1 f2 FROM t2 src ON DUPLICATE KEY UPDATE f1 = 10 + src.f1;
         select * from t2;
```
Resutlt in mysql is `11	2 ;12	1`
While in tidb-master is `11	1; 12	2`

Comment (eurekaka) [reply expression/schema.go]: Could you please help fix this as well?

Comment (jingyugao) [reply expression/schema.go]:
```sql
INSERT INTO t2 (f1, f2)
   SELECT src.f1 f2, src.f2 f1 FROM t2 src where src.f2=1
  ON DUPLICATE KEY UPDATE f1 = 100 + src.f2
```
This resolver is so strange.
There are three `src.f2`, first and second are same.But the second is not.
I will try to fix it.

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -1835,11 +1839,11 @@ func (b *PlanBuilder) buildSelectPlanOfInsert(ctx context.Context, insert *ast.I
 	schema4NewRow := expression.NewSchema(make([]*expression.Column, len(insertPlan.Table.Cols()))...)
 	for i, selCol := range insertPlan.SelectPlan.Schema().Columns {
 		ordinal := affectedValuesCols[i].Offset
-		schema4NewRow.Columns[ordinal] = &expression.Column{}
-		*schema4NewRow.Columns[ordinal] = *selCol
+		schema4NewRow.Columns[ordinal] = selCol.Clone().(*expression.Column)
+		schema4NewRow.Columns[ordinal].ColName = schema4NewRow.Columns[ordinal].OrigColName
+
+		schema4NewRow.Columns[ordinal].RetType = affectedValuesCols[i].FieldType.Clone()
```
Comment (eurekaka) [inline planner/core/planbuilder.go]: Why do we need to clone the `RetType`?

File: expression/column.go [hunk 3]
```diff
@@ -199,6 +199,8 @@ type Column struct {
 
 	hashcode []byte
 
+	// IsInsertSelect means if this column is the select field of insertSelect
+	IsInsertSelect bool
```
Comment (eurekaka) [inline expression/column.go]: This change looks too ad-hoc, we are trying very hard now to remove unnecessary fields from `Column`, do we have any other solution which is more graceful to achieve the purpose of this PR? For example, add a flag parameter in `FindColumn` to indicate if "ambiguous column" error should be ignored? @winoros Any suggestion?

Comment (jingyugao) [reply expression/column.go]: This is not easy to do.If two colums both come from `select` of `insert ... select...`,we can ignore the `ambiguous error` .But if one comes from `select` and another comes from `insert`. we must catche the error.
For example:
1. ` insert into t(a,b,f1) select f1,f2,f2 from t2 on duplicate key update a = f2` 
2. ` insert into t(a,b,f1) select f1,f2,f2 from t2 on duplicate key update a = f1`
The first one is ok, but second one is not.
So we must know each column if it is from `select` of `insert... select`.

Comment (winoros) [reply expression/column.go]: @jingyugao For you first case. I constructed another SQL
```
insert into t3(a, b, c) select f2 f1, f2 f1,f3 f1 from t2 src on duplicate key update a = f1+100;
```
Which `f1` in `a=f1+100` points to?

**Comment by winoros** [comment]
```
mysql> select * from t2;
+------+------+------+
| f1   | f2   | f3   |
+------+------+------+
|    1 |    2 |    3 |
+------+------+------+
1 row in set (0.00 sec)
mysql> select * from t3;
+---+------+------+
| a | b    | c    |
+---+------+------+
| 3 |    2 |    1 |
| 4 |    2 |    1 |
+---+------+------+
3 rows in set (0.00 sec)
mysql> insert into t3(a, b, c) select f3 f1, f2 f1,f1 xx from t2 src on duplicate key update a = a+f1;
ERROR 1062 (23000): Duplicate entry '4' for key 'PRIMARY'
mysql> insert into t3(a, b, c) select x1.f3 f1, x2.f2 f1,x1.f2 xx from t2 x1, t2 x2 on duplicate key update a = a+f1;
ERROR 1052 (23000): Column 'f1' in field list is ambiguous
```

MySQL's behavior is too strange. I think current TiDB's behavior is more intuitive. And this is the same with PG's. I would like to mark this issue as wont-fix.

**Comment by jingyugao** [comment]
> ```
> mysql> select * from t2;
> +------+------+------+
> | f1   | f2   | f3   |
> +------+------+------+
> |    1 |    2 |    3 |
> +------+------+------+
> 1 row in set (0.00 sec)
> mysql> select * from t3;
> +---+------+------+
> | a | b    | c    |
> +---+------+------+
> | 3 |    2 |    1 |
> | 4 |    2 |    1 |
> +---+------+------+
> 3 rows in set (0.00 sec)
> mysql> insert into t3(a, b, c) select f3 f1, f2 f1,f1 xx from t2 src on duplicate key update a = a+f1;
> ERROR 1062 (23000): Duplicate entry '4' for key 'PRIMARY'
> mysql> insert into t3(a, b, c) select x1.f3 f1, x2.f2 f1,x1.f2 xx from t2 x1, t2 x2 on duplicate key update a = a+f1;
> ERROR 1052 (23000): Column 'f1' in field list is ambiguous
> ```
> 
> 
> MySQL's behavior is too strange. I think current TiDB's behavior is more intuitive. And TiDB's is the same with PG's. I would like to mark this issue as wont-fix.

Maybe this is a good idea.Mysql is very different from postgresql at this grammer.
Tidb now is similar to pgsql.But now completely consistant.

**Comment by jingyugao** [comment]
@winoros 
```INSERT INTO t3 ( A ) SELECT f1 FROM t2  ON CONFLICT (A) DO UPDATE SET A = 2+f1```
```INSERT INTO t3 ( A ) SELECT f1 FROM t2  ON CONFLICT (A) DO UPDATE SET A = 2+A```
Both sql is incorrect in postgresql,pg forbidden the expr of update assignion to refer any column.
But in TiDB  it could refer any column in t3 and selectFields column in t2.

**Comment by eurekaka** [comment]
@jingyugao May I ask why is this PR closed?

**Comment by jingyugao** [comment]
> @jingyugao May I ask why is this PR closed?

Emm, I thought you dont accpet this pr.

**Pull request closed without merging**